app/src/server/api/routers/blackmarket.ts

```diff
@@ -527,3 +622,82 @@ export const fetchActiveUserOffers = async (client: DrizzleClient, userId: strin
     where: and(eq(ryoTrade.creatorUserId, userId), isNull(ryoTrade.purchaserUserId)),
   });
 };
+
+// Helper function to get rolled elements from actionLog
+async function getRolledElements(client: DrizzleClient, userId: string, elementType?: "primary" | "secondary"): Promise<{ primary: string[], secondary: string[] }> {
+  // Helper function to get rolled elements for a specific type
+  async function getRolledElementsForType(type: "primary" | "secondary"): Promise<string[]> {
```

Comment:
This entire method effectively runs up to 4 queries if both primary and secondary element are specified. I believe even in that case we can run only a single query, and then split out the informations subsequently.